File: infra/scripts/generate_supabase_jwt.py
#!/usr/bin/env python3
"""
Script para gerar JWT tokens para autenticação com PostgREST/Supabase self-hosted.

O PostgREST requer JWT tokens válidos no header Authorization: Bearer,
mas o N8N Supabase plugin espera usar o service_key diretamente.

Este script gera JWT tokens válidos que podem ser usados no lugar do service_key.
"""

import jwt
import sys
from datetime import datetime, timedelta

def generate_jwt_token(secret: str, role: str = "service_role", expires_in_days: int = 365):
    """
    Gera um JWT token válido para PostgREST.
    
    Args:
        secret: O JWT secret (service_key)
        role: O role do PostgreSQL a ser usado (anon, authenticated, service_role)
        expires_in_days: Quantos dias até o token expirar
    
    Returns:
        JWT token como string
    """
    # Payload do JWT conforme esperado pelo PostgREST
    payload = {
        "role": role,
        "exp": datetime.utcnow() + timedelta(days=expires_in_days),
        "iat": datetime.utcnow()
    }
    
    # Gerar o token usando HS256 (HMAC SHA-256)
    token = jwt.encode(payload, secret, algorithm="HS256")
    
    return token

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Uso: python3 generate_supabase_jwt.py <service_key> [role] [expires_in_days]")
        print("\nExemplo:")
        print("  python3 generate_supabase_jwt.py 'n78oYSAI5XiVxH5Ua4CYf4W+q1cS/QuSsbH9moX2onY='")
        print("  python3 generate_supabase_jwt.py 'n78oYSAI5XiVxH5Ua4CYf4W+q1cS/QuSsbH9moX2onY=' service_role 365")
        sys.exit(1)
    
    secret = sys.argv[1]
    role = sys.argv[2] if len(sys.argv) > 2 else "service_role"
    expires_in_days = int(sys.argv[3]) if len(sys.argv) > 3 else 365
    
    token = generate_jwt_token(secret, role, expires_in_days)
    
    print(f"\nJWT Token gerado para role '{role}':")
    print(f"{token}\n")
    print("Use este token no campo 'Service Role Secret' do N8N Supabase.")
    print(f"Token expira em {expires_in_days} dias.\n")

File: infra/scripts/generate_supabase_jwt.js
#!/usr/bin/env node
/**
 * Script para gerar JWT tokens para autenticação com PostgREST/Supabase self-hosted.
 * 
 * O PostgREST requer JWT tokens válidos no header Authorization: Bearer,
 * mas o N8N Supabase plugin espera usar o service_key diretamente.
 * 
 * Este script gera JWT tokens válidos que podem ser usados no lugar do service_key.
 */

const crypto = require('crypto');

function base64UrlEncode(str) {
  return str.toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=/g, '');
}

function generateJWT(secret, role = 'service_role', expiresInDays = 365) {
  const header = {
    alg: 'HS256',
    typ: 'JWT'
  };

  const now = Math.floor(Date.now() / 1000);
  const payload = {
    role: role,
    exp: now + (expiresInDays * 24 * 60 * 60),
    iat: now
  };

  const encodedHeader = base64UrlEncode(JSON.stringify(header));
  const encodedPayload = base64UrlEncode(JSON.stringify(payload));

  const signature = crypto
    .createHmac('sha256', secret)
    .update(`${encodedHeader}.${encodedPayload}`)
    .digest('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=/g, '');

  return `${encodedHeader}.${encodedPayload}.${signature}`;
}

// Main
const args = process.argv.slice(2);

if (args.length < 1) {
  console.log('Uso: node generate_supabase_jwt.js <service_key> [role] [expires_in_days]');
  console.log('\nExemplo:');
  console.log("  node generate_supabase_jwt.js 'n78oYSAI5XiVxH5Ua4CYf4W+q1cS/QuSsbH9moX2onY='");
  console.log("  node generate_supabase_jwt.js 'n78oYSAI5XiVxH5Ua4CYf4W+q1cS/QuSsbH9moX2onY=' service_role 365");
  process.exit(1);
}

const secret = args[0];
const role = args[1] || 'service_role';
const expiresInDays = parseInt(args[2]) || 365;

const token = generateJWT(secret, role, expiresInDays);

console.log(`\nJWT Token gerado para role '${role}':`);
console.log(token);
console.log(`\nTamanho: ${token.length} caracteres`);
console.log(`Token expira em ${expiresInDays} dias.`);
console.log('\nUse este token no campo "Service Role Secret" do N8N Supabase.\n');

